File: graph/sol/is-graph-bipartite.cpp
class Solution {
public:
    bool fun(vector<vector<int>>& adj , int node , int c ,vector<int>& colors){
        colors[node] = c;
        for(int v : adj[node]){
            if(colors[v] == -1){
                if(!fun(adj , v , (c^1) , colors)) return false;
            }else{
                if(colors[v] == c) return false;
            }
        }
        return true;
    }
    bool isBipartite(vector<vector<int>>& adj) {
        int N = adj.size();
        vector<int> colors(N,-1);
        for(int i = 0;i < N; i++){
            if(colors[i] == -1){
                if(!fun(adj , i , 0,colors)) return false; ;
            }
        }
        return true;
    }
};

File: graph/sol/is-graph-bipartite.java
class Solution {
    boolean dfs(int node , int c, int[] col , int[][] adj){
        col[node] = c;

        for(int neigh : adj[node]){
            if(col[neigh] == -1){
                if(!dfs(neigh , c^1 , col , adj)) return false;
            }
            if(col[neigh] == col[node]) return false;
        }
        return true;
    }
    public boolean isBipartite(int[][] adj) {
        int n = adj.length;
        int[] col = new int[n];
        Arrays.fill(col,-1);

        for(int node = 0; node < n; node++){
            if(col[node] == -1){
                if(!dfs(node,0,col,adj)) return false;
            }
        }

        return true;
    }
}
